### PR TITLE
Login modal

### LINKED ISSUE #2866: OAuth user interface design on login screen
@SidharthBansal - i wanted to share how Oauth looks on some other sites, so once we confirm it's perfectly working, we have a good workflow for login. See this example:

screen shot 2018-06-20 at 9 42 43 am

Note a couple things. First, since OAuth is easier, it's presented first. Then, I wonder if we can have the remainder of the non-OAuth screen be hidden until you interact with the first input... so as to keep things simple. Once you click or type in it, then the rest is shown. That way also it's clear that you can ignore it if you're using OAuth. 

What do you think?

### PR DESCRIPTION
**Pull request by JonathanXu1**
Fixes #2866

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` for help, in a comment below

This creates a modal that toggles when the login button is created. Email login can be toggled by email icon.

### CODE REVIEW

**Comment by oorjitchowdhary** [comment]
@jonathanxu1 I don't get the purpose of this file in your PR.. `app/views/dashboard/_comment_moderate.html.erb` .. Are you also trying to do #4048 ?

**Comment by SidharthBansal** [comment]
Can you please rebase and then push your changes?

**Comment by JonathanXu1** [comment]
Yep, I've rebased and pushed.

**Comment by SidharthBansal** [comment]
Let's wait for the tests to pass.

File: app/views/layouts/_social_icons.html.erb [hunk 1]
```diff
@@ -18,3 +18,11 @@
 		<i class="fa fa-facebook fa-fw" style="font-size:20px;color:white;"></i>
 	</span>
 </a>
+<a>
+	<span class="btn btn-default" style="margin-left:2px; background-color:#7c7c7c; color:white; height:36px; " id="toggle-btn">
+		Log in with email
+	</span>
+</a>
+
+<% if Rails.env == "production" %>
```
Comment (SidharthBansal) [inline app/views/layouts/_social_icons.html.erb]: Please remove line 27 and 28

File: app/views/dashboard/_comment_moderate.html.erb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<% if comment.status == 4 %>
```
Comment (SidharthBansal) [inline app/views/dashboard/_comment_moderate.html.erb]: I think these changes are not related to the issue upon which you are working upon. Can you please remove them?

File: app/views/tag/_sorting.html.erb [hunk 3]
```diff
@@ -7,7 +7,11 @@ line up nicely if the user is asked to login on the questions/index page -->
 
 <form id="tagform" class="form-horizontal" method="GET" action="/questions/tag/">
   <div class="input-group">
+<<<<<<< HEAD
```
Comment (SidharthBansal) [inline app/views/tag/_sorting.html.erb]: I think you have not completely resolved the conflicts.
Either you can create a new branch with the updated branch and commit your changes in a new pr and close this one
OR you can rebase this pr and remove all the conflicts.

Please remove these changes.

**Comment by JonathanXu1** [comment]
@SidharthBansal resolved those issues, pushing now...

**Comment by SidharthBansal** [comment]
You were initially talking about some issues which you are facing for this issue.
Are they now resolved?
Please provide a final screenshot before we can merge this.
Thanks a lot

**Comment by SidharthBansal** [comment]
Once this is complete, you need to create another PR for integrating the sign-up form(pl.org/signup) in the modal so that we will have a single page for login/signup form.
As GCI tasks are small, so sign up will be in another pr.
Please create an issue for it also and link it to this pr.
Thanks

**Comment by SidharthBansal** [comment]
@JonathanXu1 all the login/signup things will go into OAuth milestone. So if you find any issue on which you can add the milestone please add it.
Thanks

**Comment by JonathanXu1** [comment]
Yep, the initial issues you mentioned are now resolved. Here's a screenshot:
image
Can I claim this submit this task on gci and create a new issue for adding signup?

**Comment by JonathanXu1** [comment]
@SidharthBansal here's the issue I've just created for signup: #4173

**Comment by SidharthBansal** [comment]
Last thing, just submit a gif so that we can be sure that 
1. Sign Up works for modal
2. Modal is able to be open up and closed by clicking on 'X' button on the rightmost topmost corner of the modal.

**Comment by JonathanXu1** [comment]
Here's the gif file demonstrating:

- x button works
- login by email can be toggled with the respective button
- signup modal works on only on the signup page, when I go to any url other than /signup it breaks
ezgif com-video-to-gif

**Comment by jywarren** [comment]
WHOA this is already done? 

wow!

OK, well I'm just being extra cautious because signup/login is such a critical portion. We could still merge in ONLY the login modal code for the time being, if you think it's ready. I would like to see the signup modal at different page widths -- mobile view, etc. -- just to be sure it works there too!

**Comment by JonathanXu1** [comment]
@jywarren yeah I think just merging login is a much safer and user-friendly option. Signup is very iffy as it only works on the signup page and is very large as a modal. I'll submit my task once I apply the changes to the login modal.

**Comment by SidharthBansal** [comment]
Sure

**Comment by jywarren** [comment]
OK great. Thanks, everyone! Good collective decisionmaking! :-)

**Comment by SidharthBansal** [comment]
Can you please center the modal to the screen?

**Comment by JonathanXu1** [comment]
How is this? 
image

**Comment by SidharthBansal** [comment]
@JonathanXu1 have you removed the changes for the sign up modal from this pr?

**Comment by SidharthBansal** [comment]
Please provide one more screenshot when the username, email address field is visible.
Changes looks good to me.

**Comment by JonathanXu1** [comment]
@SidharthBansal toggle button is now removed.

image

**Comment by jywarren** [comment]
This is beautiful!!!!! ⚡🙌🏼❤️

File: app/views/users/_form.html.erb [hunk 4]
```diff
@@ -8,68 +8,67 @@
...
-      <div class='col-lg-10 col-md-10 col-sm-10' style="display:flex; justify-content: center;">
-        <%= render :partial => "layouts/social_icons" %>
-      </div>
-
-   <div class='col-lg-10 col-md-10 col-sm-10' style="display:flex; justify-content: center;">
-    <h3> OR</h3>
 
-   </div>
-   <br style="clear:both;"/>
+    <div class='col-lg-10 col-md-10 col-sm-10' style="display:flex; justify-content: center;">
```
Comment (jywarren) [inline app/views/users/_form.html.erb]: Ah i see there are still some changes here, but are you confident the form still works?

**Comment by JonathanXu1** [comment]
I can create a test account, and there are no errors in the console. As far as I can tell, the form works fine. The center align div was added while attempting to work on a signup modal but that should have no effect.
image

File: app/views/layouts/_social_icons.html.erb [hunk 5]
```diff
@@ -1,20 +1,46 @@
...
+			</span>
+		</a>
+		<a href="/auth/facebook?origin=<%= params[:return_to] || root_url %>" id="connect-facebook">
+			<span class="btn btn-default" style="margin-left:2px;background-color: #3b5998">
+				<i class="fa fa-facebook fa-fw" style="font-size:20px;color:white;"></i>
+			</span>
+		</a>
+	</div>
+	
+	<div class='hr-or'></div>
```
Comment (SidharthBansal) [inline app/views/layouts/_social_icons.html.erb]: Actually @JonathanXu1 has shifted ``OR`` in the _social_icons partial.
So, there are some changes for the sign up form.
We can merge this, I think. Also, the screenshot depicts well the changes does not effect sign up form.

**Comment by SidharthBansal** [comment]
Yeah, we @jywarren as you have approved the changes and I have also reviewed it, I have merged it.
Thanks @JonathanXu1 can you please claim the issue on GCI dashboard so that we can give you points?

**Comment by oorjitchowdhary** [comment]
So the sign up is ready too? Sorry, I didn't quite follow the conversation

**Comment by SidharthBansal** [comment]
@JonathanXu1 is facing some issue in the Sign up, so it will be great if we can provide him some help there.

**Comment by SidharthBansal** [comment]
@oorjitchowdhary we will be processing 3 prs, 
1.Login modal
2.Sign Up Modal
3.Integrating above 2 into 1

**Comment by oorjitchowdhary** [comment]
Okay.. Sure we'll help @JonathanXu1

**Comment by oorjitchowdhary** [comment]
I'm unable to log in on localhost through the modal..
The password field doesn't work when clicked..
I'll make a gif to demonstrate..

Moreover, all the oauth aren't working... They lead to `404 error`

**Comment by SidharthBansal** [comment]
The oauth providers will show 404 error on local host as the keys are not set on your laptop. Don't raise an issue for it.
You can ask Jonathan for a fixed for the password field.

**Comment by oorjitchowdhary** [comment]
Okay

**Comment by SidharthBansal** [comment]
@oorjitchowdhary the login button for the modal is not working.
Else everything is working smoothly. All Oauth providers are working.
If you wish you can send a patch for it while @JonathanXu1 is away from the computer.
I have pushed the changes to the unstable

**Comment by oorjitchowdhary** [comment]
Oh. What is the error.. Is it some error on 'POST'??

**Comment by SidharthBansal** [comment]
Most probably the error is POST.
The form is trying to post but cannot.
I guess we should try f.submit for form submission in modal instead of submit.

**Comment by SidharthBansal** [comment]
Reporting error counts as an issue solved. Please claim the required task so that I can reward you.

**Comment by SidharthBansal** [comment]
@oorjitchowdhary can you also check the current master on the localhost.
Behaviour to be checked: A user enters wrong email and password.
Expected Output: Modal should display the errors. Login form should also display the errors.

**Comment by SidharthBansal** [comment]
@oorjitchowdhary just check the last comment for login. Other things I have checked
Thanks a lot

**Comment by SidharthBansal** [comment]
https://github.com/publiclab/plots2/pull/4203#issuecomment-445546816
Awesome.
Please claim the task so that I can give you rewards.

**Comment by SidharthBansal** [comment]
You need to take sign up modal task on th GCI platform.

**Comment by oorjitchowdhary** [comment]
Okay..

**Comment by oorjitchowdhary** [comment]
Okay.. I've claimed it

**Comment by SidharthBansal** [comment]
Moved to #4210
